neuralcompression/models/_unet_discriminator.py

```diff
@@ -0,0 +1,337 @@
+
+
+def apply_conv_norm(module: nn.Module, norm_type: str = "identity") -> nn.Module:
+    _verify_norm_type(norm_type)
+    if norm_type == "spectral":
+        return spectral_norm(module)
+    else:
+        return module
+
+
+def build_norm_layer(chans: int, norm_type: str) -> nn.Module:
+    _verify_norm_type(norm_type)
+    if norm_type == "batch":
+        return nn.BatchNorm2d(chans)
+    elif norm_type == "instance":
+        return nn.InstanceNorm2d(chans)
+    else:
+        return nn.Identity()
+
+
+def _conv(
+    in_planes: int,
+    out_planes: int,
+    kernel_size: int = 3,
+    stride: int = 1,
+    padding: int = 1,
+    bias: bool = True,
+):
+    return nn.Conv2d(
+        in_planes,
+        out_planes,
+        kernel_size=kernel_size,
+        stride=stride,
+        padding=padding,
+        bias=bias,
+    )
+
+
+class _UnetLevel(nn.Module):
+    def __init__(
+        self,
+        in_planes: int,
+        down_planes: int,
+        child: nn.Module,
+        out_planes: Optional[int] = None,
+        top: bool = False,
+        clip_upsample: bool = False,
+        norm_type: str = "identity",
+    ):
+        super().__init__()
+        self.in_planes = in_planes
+        self.down_planes = down_planes
+        if out_planes is None:
+            out_planes = in_planes
+        self.out_planes = out_planes
+        self.child = child
+        self.clip_upsample = clip_upsample
+        self.norm_type = norm_type
+
+        use_bias = True
+        if norm_type == "batch":
+            use_bias = False
+
+        norm_layer = partial(build_norm_layer, norm_type=norm_type)
+        conv_norm = partial(apply_conv_norm, norm_type=norm_type)
+
+        self.left_block = nn.Sequential(
+            nn.Identity() if top else norm_layer(in_planes),
+            nn.Identity() if top else nn.LeakyReLU(negative_slope=0.2, inplace=True),
+            conv_norm(
+                _conv(in_planes=in_planes, out_planes=down_planes, bias=use_bias)
+            ),
+            norm_layer(down_planes),
+            nn.LeakyReLU(negative_slope=0.2, inplace=True),
+            conv_norm(_conv(in_planes=down_planes, out_planes=down_planes)),
+            nn.AvgPool2d(2),
+        )
+        if in_planes != down_planes:
+            left_shortcut_conv = conv_norm(
+                _conv(
+                    in_planes=in_planes,
+                    out_planes=down_planes,
+                    kernel_size=1,
+                    padding=0,
+                    bias=True,
+                )
+            )
+        else:
+            left_shortcut_conv = nn.Identity()
+
+        self.left_shortcut = nn.Sequential(
+            left_shortcut_conv,
+            nn.AvgPool2d(2),
+        )
+
+        self.right_block: nn.Module
+        self.right_shortcut: Union[nn.Module, Callable]
+        if clip_upsample is True:
+            self.right_block = nn.Identity()
+            self.right_shortcut = lambda x: 0.0
```

Comment:
should it return instead a torch.tensor ? I guess it will be automatically casted to it anyway ?